### ISSUE TITLE
Override load-balancer healthcheck target via annotations

### ISSUE THREAD

**Issue by bordeuax** [kubernetes/cloud-provider-aws#49]
**What would you like to be added**:  
Have a kubernetes service annotation like "service.beta.kubernetes.io/aws-load-balancer-healthcheck-target", that can override the default healthcheck target and follow the format from aws-sdk for Go lang , for more info see  Target field  https://docs.aws.amazon.com/sdk-for-go/api/service/elb/#HealthCheck.
   

**Why is this needed**:
I have two uses cases 
1.  One of our component require client certificate authentication and all the traffic is handled by LB as TCP ,  HealthCheck from LB doing simple TCP connections  that generate a lot of logs with SSL errors in ours pods. To avoid this it's will be nice to deploy the K8S service with : 

```
apiVersion: v1
kind: Service
metadata:
  labels:
    project: 
  annotations:
       service.beta.kubernetes.io/aws-load-balancer-healthcheck-target:  "SSL:3356"
```
if the port is not specifyed then by default is from application 
```
apiVersion: v1
kind: Service
metadata:
  labels:
    project: 
  annotations:
       service.beta.kubernetes.io/aws-load-balancer-healthcheck-target:  "SSL"
```

2.  The healthcheck of application is accessible via different port(comparing with applicaiton port) and have a specific path, in this case the solution can be : 

```
apiVersion: v1
kind: Service
metadata:
  labels:
    project: 
  annotations:
     service.beta.kubernetes.io/aws-load-balancer-healthcheck-target:  "HTTP:3636/app/status/lb"

```

/kind feature
/remove-lifecycle stale

**Comment by vvbogdanov87** [comment]
Kuberntes 1.14 [uses SSL health checks for ELBs when backend protocol is SSL/HTTPS](https://github.com/kubernetes/kubernetes/pull/70309)
But for the health check pIng target k8s selects first port in the listeners list
https://github.com/2rs2ts/kubernetes/blob/72895a84a9670d6f6f8921681c5bbe4b2745319e/pkg/cloudprovider/providers/aws/aws.go#L3602
Without annotation described above, you MUST set https/SSL port to be first in the `spec.ports`
I had a service spec:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: myapp
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: https
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: https
spec:
  type: LoadBalancer
  selector:
    app: myapp
  ports:
    - name: http
      protocol: TCP
      port: 80
    - name: https
      protocol: TCP
      port: 443
```
After I upgraded to k8s 1.14 I got `OutOfService` for all instances in AWS LB. I had to change the spec to
```
  ports:
    - name: https
      protocol: TCP
      port: 443
    - name: http
      protocol: TCP
      port: 80
```

**Comment by nicolasbelanger** [comment]
Got burned by this too today.  Consequence:  no more ingress, useless cluster.  We're using ambassador as the gateway, and the default values for the service is exactly as @vvbogdanov87 described above, http first, https second.  SSL ping target was done on the http port, which failed.

**Comment by 2rs2ts** [comment]
Use case 2 in the original post is pretty common at my company, so I'd love to have this. We do not use the standalone provider at my company though so I don't think even if I could find the time to implement this that I would know how to test it.

**Comment by afnanenayet** [comment]
We ran into this today as well at my company. I believe that in GKE the health check from the deployment propagates to the google cloud ELB, so you don't even need to set annotations if you want parity between the external load balancer and your internal kube service.

**Comment by steerben** [comment]
I think that is indeed a very valuable feature.

Considering subsequent use case:

HTTPS (ELB - ACM SSL termination) -> HTTPS (Worker Self signed SSL termination)

In that case the healthcheck also turns into HTTPS. However most health check endpoints run on separate ports and do not support SSL termination like e.g. the one of the istio ingress gateway which renders the mentioned use case not feasible at the moment.

**Comment by kishorj** [comment]
Supports available for specifying https protocol for NLB instance mode health check via annotatoin `service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol`  https://github.com/kubernetes/kubernetes/issues/92321. Fix will be included in 1.20.

NLB IP mode also supports the annotation, and is available in the [AWS Load Balancer controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller)

**Comment by nckturner** [comment]
/close

Ref:
https://github.com/kubernetes/kubernetes/issues/92321
https://github.com/kubernetes/kubernetes/pull/94546

This fix is included in the external cloud controller manager (this repository) as well.
